### ISSUE TITLE
Optional different formatting for output table for microtesting

### ISSUE THREAD

**Issue by danielloader** [nils-braun/pytest-benchmark-commenter#6]
Just want to start by saying thanks for creating this, it's a great addition to the github actions market.

Just deployed it to test our new library, but the library is a log parser where the tests are doing tight loops on the parsing methods, so the attributes we care most about are: operations per second, rounds taken and mean. However the mean times are in microseconds, so round down to `0.00` on the current table. 

Not sure if you'd be interested in a fully configurable way to decide on units, and which columns to include, or just have a templated format option between normal testing and this sort of testing. 

```
----------------------------------------------------------------------------------------------------- benchmark: 8 tests -----------------------------------------------------------------------------------------------------
Name (time in us)                                     Min                   Max                Mean              StdDev              Median                IQR            Outliers  OPS (Kops/s)            Rounds  Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_apache_access_speed                          19.3000 (1.0)        370.1000 (1.93)      21.7065 (1.0)        6.0220 (1.0)       20.3000 (1.0)       0.7000 (1.00)    1432;2866       46.0692 (1.0)       21979           1
test_cloudflare_access_speed                      48.8000 (2.53)       352.4000 (1.84)      54.0498 (2.49)      14.2508 (2.37)      50.6000 (2.49)      1.5000 (2.14)     536;1256       18.5014 (0.40)       8937           1
test_apache_access_speed_validation              102.7000 (5.32)       191.6000 (1.0)      107.8831 (4.97)      10.5300 (1.75)     105.0000 (5.17)      1.4500 (2.07)        10;22        9.2693 (0.20)        148           1
test_cloudflare_access_speed_with_validation     145.4000 (7.53)     3,078.3000 (16.07)    292.0481 (13.45)    265.1503 (44.03)    240.7500 (11.86)    56.1000 (80.14)        7;20        3.4241 (0.07)        158           1
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Comment by nils-braun** [comment]
Hi @danielloader!

Thank you for using the package! I am keen to learn about your project as well :-)

I like your idea and I can definitely see your use case. 
I think we can have a fully configurable option, maybe with some examples on the documentation page. Would you think it is enough to give a dict-like structure of column names and units (maybe with default units?) or is more information needed?

**Comment by danielloader** [comment]
Going to give it a go to come up with some prototype even if TS isn't my background, though the prototype will probably have to omit comparison to previous HEAD for brevity of proving value in the idea.

**Comment by nils-braun** [comment]
That would be super nice! If you need any help, do not hesitate to get in touch with me!

**Comment by danielloader** [comment]
Haven't raised a PR cross repo as it's clearly a very rough first draft and subject to some wild changes however...
https://github.com/danielloader/pytest-benchmark-commenter/pull/2

Firstly I've grabbed an off the shelf object to markdown table library, too many conditional outcomes to be fully dynamic with string splicing.

![image](https://user-images.githubusercontent.com/933423/125487873-6446f630-af4c-45c9-b243-791b1d967e0f.png)
This is how it's looking with no comparison data, the comparison example is in the PR above using some slightly modified data (since the ops count in both was the same).

Very open to how we can take this forward, I'm not usually a fan of adding too many configuration options but I can't see a way to reduce them safely.

To summarise the changes:
- Configurable names for benchmarks (sometimes you might just want the short name rather than the fullname, useful when you're trying to display more columns)
- Configurable time unit scaling for second floats
- Configurable comparison metric rather than just mean
- Configurable metrics to show in table

PS: Typescript is interesting if you only know Python.

**Comment by danielloader** [comment]
Further things to consider:
- Auto unit clamping, setting the time unit to "auto" so it can work out how many proceeding 0s in the significant figures it goes through on the float until something useful comes up:
1.011 = seconds = 1.01s
0.00132 = milliseconds = 1.32ms

- Scaling the ops value like pyest, when the number is large enough make it kiloOps, megaOps etc.

**Comment by danielloader** [comment]
Okay, taken my ideas off the previous comment and gave it a stab: https://github.com/danielloader/pytest-benchmark-commenter/pull/3

- Unit types have some sort of unit scaling now (per benchmark per dataset).
- Removed the name vs fullname and just rolled it into the same yaml for defining columns.
- Removed time unit configuration.

Example configuration below produces the following table:
```
- name: Run the action
  uses: ./ # Uses an action in the root directory
  with:
    benchmark-file: "output.json"
    benchmark-metrics: 'name,max,min,mean,ops'
    comparison-benchmark-file: "old_benchmark/output.json"
    comparison-benchmark-metric: "mean"
```

![image](https://user-images.githubusercontent.com/933423/125648270-956b2764-d931-437c-bb7a-7a43364c105a.png)

Second stab at a refactor, learning the ropes with typescript as I go, hopefully it sparks some ideas!

**Comment by nils-braun** [comment]
Hi @danielloader 
That looks already super cool! Just wanted to say I did not forget about you issue, I am just a little busy currently. I will get back with an answer in the next days. Already thank you for your changes and time!
Did you really learn typescript while doing this? Impressive!

**Comment by nils-braun** [comment]
I have finally found the time to have a look into your changes. I really like them. You did not only add new functionality, but you also cleaned up the code (which was long overdue). Thank you so much! I think the new structure is really good. Feel free to open a "real" PR if you want - I do not think I will have many comments while doing the review (except maybe some documentation comments here and there and a change of the README. However I am happy to support you here).

I have also thought about the configuration. One other possibility is to ask the user to give a "dictionary" like

```yaml
column-config:
  - name: Standard Deviation
    source: std
    unit: auto # could be the default
  - name: My fancy name # can be chosen differently
    source: min
    unit: ms 
  - name: Maximum
    source: max
    unit: auto
    show_change: true # would make it show up in the comparison with HEAD
```

That would give all options to the user but I just feel that this is just too much possibilities. I can not think of a reasonable use-case where you would like to change the titles and/or the units from auto for the columns differently. The only slight benefit would be, that the tool might also work for non-pytest-benchmark-files then, but I think it would be better to just write another action in this case... Long story short: let's stay with your solution!

**Comment by danielloader** [comment]
Hey, thanks for reviewing it preliminarily. I too had the same thought initially added way too many configurable toggles and it became unweildly, having which columns to display and which one to compare on seems "enough" for now at least.

Going forth it might be worth splitting it into two tables, or two rows per test. Shame github markdown can't do font colouring for better or worse metrics...

Do you want to open a branch off master and I do a PR into that branch so it's in your hands or a PR against master and then do the changes inline in the PR? As you say readme changes need doing at the least. Think it's probably smart to include some pytest flag suggestions too for micro benchmarking, like the increasing minimum iterations could be beneficial.

As to learning Typescript, yeah was a bit of a trial by _fire_! 

It's not the cleanest, but I felt exposing the same methods in all class objects aided "normalisation" when displaying in text at the render phase. Played with some other methods but it felt the sanest and easiest to extend with different object to string renderers.

**Comment by nils-braun** [comment]
Hi @danielloader!
Feel free to open a PR against master from your fork. That is the easiest solution and you still "own" your branch :-) I will read the PR once open and maybe have some suggestions.

I like having a better overview - but I think I did not quite understand your proposal with the two tables. Would you like to create one table for the HEAD and one for the branch?

**Comment by danielloader** [comment]
Potentially, it's the logical answer if you need to compare multiple metrics, but for now it's completely fine to have a single point of comparison per test suite, gives you the capacity to split the tests up if need be and run pytest repeatedly with different output.json files and render them differently. 

Will raise it soon.
